tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/command_queue/test_EnqueueQueueWaitForEvent.cpp

```diff
@@ -53,5 +53,81 @@ TEST_F(MultiCommandQueueSingleDeviceFixture, TestEventsEventSynchronizeSanity) {
+            EXPECT_EQ(event.event_id, j * 2);
+            EnqueueQueueWaitForEvent(cqs[i], event);
+        }
+    }
+    local_test_functions::FinishAllCqs(cqs);
+}
+
+// Record event on one CQ, wait-for-that-event on another CQ. Then do the flip. Occasionally insert
+// syncs from Host per CQ, and verify completion queues per CQ are correct.
+TEST_F(MultiCommandQueueSingleDeviceFixture, TestEventsEnqueueQueueWaitForEventCrossCQs) {
+    CommandQueue a(this->device_, 0);
+    CommandQueue b(this->device_, 1);
+    vector<std::reference_wrapper<CommandQueue>> cqs = {a, b};
+    size_t num_events_per_cq = 10;
+
+    // Currently hardcoded for 2 CQ. For 3+ CQ, can extend to record for CQ0, Wait for CQ1,CQ2,etc.
+    TT_ASSERT(cqs.size() == 2);
+    const int num_cmds_per_cq = 2;
+
+    // Store completion queue base address from initial rdptr, for later readback.
+    vector<uint32_t> completion_queue_base;
+    for (uint i = 0; i < cqs.size(); i++) {
+        completion_queue_base.push_back(this->device_->sysmem_manager().get_completion_queue_read_ptr(i));
+    }
+
+    // Issue a number of Event Record/Waits per CQ, with Record/Wait on alternate CQs
+    for (size_t j = 0; j < num_events_per_cq; j++) {
+        for (uint i = 0; i < cqs.size(); i++) {
+            auto &cq_record_event = cqs[i];
+            auto &cq_wait_event = cqs[(i + 1) % cqs.size()];
+            Event event;
+            log_debug(tt::LogTest, "Recording event on CQ ID: {} and Device Syncing on CQ ID: {}", cq_record_event.get().id(), cq_wait_event.get().id());
+
+            EnqueueQueueRecordEvent(cq_record_event, event);
+            EXPECT_EQ(event.cq_id, cq_record_event.get().id());
+            EXPECT_EQ(event.event_id, i + (j * num_cmds_per_cq));
+            EnqueueQueueWaitForEvent(cq_wait_event, event);
+
+            // Occasionally do host wait for extra coverage from both CQs.
+            if (j > 0 && ((j % 3) == 0)) {
+                EventSynchronize(event);
+            }
+        }
+    }
+
+    local_test_functions::FinishAllCqs(cqs);
+
+    // Check that completion queue per device is correct. Ensure expected event_ids seen in order.
+    chip_id_t mmio_device_id = tt::Cluster::instance().get_associated_mmio_device(this->device_->id());
+    uint16_t channel = tt::Cluster::instance().get_assigned_channel_for_device(this->device_->id());
+    constexpr uint32_t completion_queue_event_alignment = 32;
+    uint32_t event;
+
+    for (uint cq_id = 0; cq_id < cqs.size(); cq_id++) {
+        for (size_t i = 0; i < num_cmds_per_cq * num_events_per_cq; i++) {
+            uint32_t host_addr = completion_queue_base[cq_id] + i * completion_queue_event_alignment;
+            tt::Cluster::instance().read_sysmem(&event, 4, host_addr, mmio_device_id, channel);
+            EXPECT_EQ(event, i);
+        }
+    }
+}
+
+// TODO/WIP - More tests, mixing reads/writes potentially and using events to synchronize Read-After-Write.
```

Comment:
tests above appear to cover well that events happen in order alone and across queues, but not clear we have a test that ensures that events wait until, eg, a kernel completes.  wonder if you could run something that runs long and returns some data and then afterwards do a host write to trash that location of L1.  if the event completes early, then the data returened be corrupt.  or maybe this is covered somehow?

Reply:
Yep, I have other tests below and newly added after this and WIP that will further test multi-cq ordering, this one was purposely kept ~simple.